mono/metadata/sgen-tarjan-bridge.c

```diff
@@ -536,28 +574,121 @@ push_object (GCObject *obj)
 		data = create_data (obj);
 	g_assert (data->state == INITIAL);
 	g_assert (data->index == -1);
+
+	*forward_obj = data;
+	*push_obj = data;
+}
+
+static void
+mark_object_scanned (ScanData *data)
+{
+	data->low_index = data->index = object_index++;
 	dyn_array_ptr_push (&scan_stack, data);
+	dyn_array_ptr_push (&loop_stack, data);
+}
+
+// The target of the scan gets pushed onto the stack first.
+// However, when using OPTIMIZATION_FORWARD, we want to avoid pushing the object
+// if it has exactly one forwardable reference. Therefore we wait to push the
+// scan target until we know this criteria has failed.
+// initial_scan_one returns as arguments a forwardable argument (if any) and
+// a pushable object (if any). pushable objects are always forwardable.
+static void
+initial_scan_one (ScanData *data, GCObject *obj, ScanData **first_push, ScanData **first_forward, gboolean *forward_check_finished)
+{
+	ScanData *to_push = NULL, *to_forward = NULL;
+	initial_scan_check(data, obj, &to_push, &to_forward);
+	if (*forward_check_finished) {
+		if (to_push) {
+			dyn_array_ptr_push (&scan_stack, to_push);
+		}
+	} else if (to_forward) {
+		if (*first_forward) {
+			mark_object_scanned(data);
+			*forward_check_finished = TRUE;
+
+			if (*first_push)
+				dyn_array_ptr_push (&scan_stack, *first_push);
+			if (to_push)
+				dyn_array_ptr_push (&scan_stack, to_push);
+		} else {
+			*first_push = to_push;
+			*first_forward = to_forward;
+		}
+	}
 }
 
 #undef HANDLE_PTR
 #define HANDLE_PTR(ptr,obj)	do {					\
 		GCObject *dst = (GCObject*)*(ptr);			\
-		if (dst) push_object (dst); 			\
+		if (dst)                                    \
+			initial_scan_one (data, dst, &first_push, &first_forward, &forward_check_finished ); \
 	} while (0)
 
 // dfs () function's queue-children-of-object operation.
 static void
-push_all (ScanData *data)
+initial_scan (ScanData *data)
 {
 	GCObject *obj = data->obj;
 	char *start = (char*)obj;
 	mword desc = sgen_obj_get_descriptor_safe (obj);
 
+	gboolean forward_check_finished = FALSE;
```

Comment:
Could you document how forwarding works in terms of forward_check_finished, first_push and first_forward?

The code is very hard to make sense of.